tsconfig.json

```diff
@@ -0,0 +1,24 @@
+{
+  "compilerOptions": {
+    "target": "ES2017",
```

Comment:
I think producing an ES2017 or even ES2018 version is a good idea. :+1:
Could we also build an ES5 version?

See: https://babeljs.io/blog/2018/06/26/on-consuming-and-publishing-es2015+-packages

Reply:
@zeusdeux I’m also interested in this one.

Reply:
@ChristianMurphy so if I understand correctly, you’d want `main` to be the entry for the code transpiled to ES5 + CJS and `module` to be code transpiled to ES2017/ES2018+ES6 modules?

Reply:
> main to be the entry for the code transpiled to ES5 + CJS 

:+1: exactly

> `module` to be code transpiled to ES2017/ES2018+ES6 modules?

Close, from the babel article:

> However, the sole intention of this field is ESM, not anything else. The Rollup docs specify that the `module` field makes it clear that it's not intended for future JavaScript syntax.
>
> Despite this warning, package authors invariably conflate the use of ES modules with the JavaScript language level they authored it in.

https://babeljs.io/blog/2018/06/26/on-consuming-and-publishing-es2015+-packages#conflating-javascript-modules-and-es2015

publishing ES2015+ under `module` can break rollup builds and create react app v1 builds.

One of the points of the articles is there is not _yet_ a standard way to ship ES2015+ in packages.
The proposed syntax right now `main-es` or `es`.
Which runs along side the existing `main` and `module` (both in ES5).

Reply:
Tom just wrote a nice post about how to bundle this all, except the typescript part: https://macwright.org/2018/11/26/small-modules.html, it mentions `microbundle` as well, just like the babel post above.

Reply:
Looks like microbundle supports Typescript out of the box now.
https://github.com/developit/microbundle/pull/16

Reply:
Nice, I’m up for trying it, what do y’all think?

Reply:
Or maybe the current setup is good enough already?

Reply:
@ChristianMurphy haha I was going to bring up the same point from the blog if you'd have said yes wrt `module`. Since there isn't consensus on that now, I'll do the following:

1. `main` will point to an entry for code that is transpiled to ES5 and uses CommonJS module system
2. `module` will point to an entry for code that is transpiled to ES5 but uses ES6 module system so that bundlers can do cool things like DCE/tree-shaking
3. `main-es` and `es` will both point to an entry for code that is transpiled to ES2017 (I don't see ES2018 in the compiler options and ESNext is too aggressive imho) and uses ES6 module system

How does that sound @ChristianMurphy @wooorm?
Also, any opinions on including a umd build as well?

Reply:
I like that. And :-1: to umd, I don’t see a lot of value for that in such a low-level project.

Reply:
Instead of adding yet another tool, I have chosen to use the typescript compiler to do this as it supports all of this by default. A commit shall follow in the next few minutes.

Reply:
The commit is blocked by https://github.com/micromark/micromark/pull/2#issuecomment-441693780 :(